cassandra/connection.py

```diff
@@ -360,6 +368,28 @@ def error_all_requests(self, exc):
                             "failed connection (%s) to host %s:",
                             id(self), self.host, exc_info=True)
 
+        # run first callback from this thread to ensure pool state before leaving
+        cb, _ = requests.popitem()[1]
+        try_callback(cb)
+
+        if not requests:
+            return
+
+        # additional requests are optionally errored from a separate thread
+        # The default callback and retry logic is fairly expensive -- we don't
+        # want to tie up the event thread when there are many requests
+        def err_all_callbacks():
+            for cb, _ in requests.values():
+                try_callback(cb)
+        if len(requests) < Connection.CALLBACK_ERR_THREAD_THRESHOLD:
+            err_all_callbacks()
+        else:
+            # daemon thread here because we want to stay decoupled from the cluster TPE
+            # TODO: would it make sense to just have a driver-global TPE?
+            t = Thread(target=err_all_callbacks)
```

Comment:
I guess another option would be a lazily-constructed TPE for connections. I don't think that's worth the complexity based on the above reasoning. Interested to hear what you think.